internal/test/container.go

```diff
@@ -140,22 +142,45 @@ func runTestWithSession(ctx context.Context, logger *log.Logger, w io.Writer, do
 		}
 
 		var dockerfileTarball bytes.Buffer
-		if err := createDockerfileTarball(tar.NewWriter(&dockerfileTarball), dockerfile); err != nil {
+		if err = createDockerfileTarball(tar.NewWriter(&dockerfileTarball), dockerfile); err != nil {
 			return err
 		}
 
-		logger.Println("creating test image")
-		imageBuildResult, err := dockerDaemon.ImageBuild(ctx, &dockerfileTarball, types.ImageBuildOptions{
-			Tags:      []string{"kiln_test_dependencies:vmware"},
-			Version:   types.BuilderBuildKit,
-			SessionID: sessionID,
-		})
-		if err != nil {
-			return fmt.Errorf("failed to build image: %w", err)
-		}
+		if configuration.ImagePath == "" {
+			logger.Println("creating test image")
+			imageBuildResult, err := dockerDaemon.ImageBuild(ctx, &dockerfileTarball, types.ImageBuildOptions{
+				Tags:      []string{"kiln_test_dependencies:vmware"},
+				Version:   types.BuilderBuildKit,
+				SessionID: sessionID,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to build image: %w", err)
+			}
+			if err = checkSSHPrivateKeyError(imageBuildResult.Body); err != nil {
+				return err
+			}
+		} else {
+			logger.Println("loading test image")
+			imageReader, err := os.Open(configuration.ImagePath)
+			if err != nil {
+				return fmt.Errorf("failed to read image '%s': %w", configuration.ImagePath, err)
+			}
 
-		if err := checkSSHPrivateKeyError(imageBuildResult.Body); err != nil {
-			return err
+			loadResponse, err := dockerDaemon.ImageLoad(
+				ctx,
+				imageReader,
+				true,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to import image: %w", err)
+			}
+
+			respBytes, err := io.ReadAll(loadResponse.Body)
+			defer loadResponse.Body.Close()
+			if err != nil {
+				return fmt.Errorf(`failed to parse load image response: %w`, err)
+			}
+			logger.Printf("loaded image %s: \n%s\n", configuration.ImagePath, string(respBytes))
```

Comment:
Please defer the close of your writer right after the error check in case io.ReadAll panics for some odd reason:

```patch
diff --git a/internal/test/container.go b/internal/test/container.go
index 8f8bfa9b..f8b44fcc 100644
--- a/internal/test/container.go
+++ b/internal/test/container.go
@@ -174,9 +174,9 @@ func runTestWithSession(ctx context.Context, logger *log.Logger, w io.Writer, do
 			if err != nil {
 				return fmt.Errorf("failed to import image: %w", err)
 			}
+			defer closeAndIgnoreError(loadResponse.Body)
 
 			respBytes, err := io.ReadAll(loadResponse.Body)
-			defer loadResponse.Body.Close()
 			if err != nil {
 				return fmt.Errorf(`failed to parse load image response: %w`, err)
 			}
```

Reply:
done